src/java/org/apache/cassandra/io/util/WrappingRebufferer.java

```diff
@@ -81,75 +75,32 @@ public void closeReader()
         source.closeReader();
     }
 
-
     @Override
     public String toString()
     {
         return String.format("%s[]:%s", getClass().getSimpleName(), source.toString());
     }
 
-    protected final class WrappingBufferHolder implements BufferHolder
+    @Override
+    public ByteBuffer buffer()
     {
-        @Nullable
-        private BufferHolder bufferHolder;
-
-        private ByteBuffer buffer;
-        private long offset;
-
-        protected WrappingBufferHolder initialize(@Nullable BufferHolder bufferHolder, ByteBuffer buffer, long offset)
-        {
-            assert this.bufferHolder == null && this.buffer == null && this.offset == 0L : "initialized before release";
-
-            this.bufferHolder = bufferHolder;
-            this.buffer = buffer;
-            this.offset = offset;
-
-            return this;
-        }
-
-        @Override
-        public ByteBuffer buffer()
-        {
-            return buffer;
-        }
-
-        @Override
-        public long offset()
-        {
-            return offset;
-        }
-
-        public int limit()
-        {
-            return buffer.limit();
-        }
-
-        public void limit(int limit)
-        {
-            this.buffer.limit(limit);
-        }
-
-        @Override
-        public void release()
-        {
-            assert buffer != null : "released twice";
-
-            if (bufferHolder != null)
-            {
-                bufferHolder.release();
-                bufferHolder = null;
-            }
-
-            buffer = null;
-            offset = 0L;
+        return buffer;
+    }
 
-            buffers.offerFirst(this);
-        }
+    @Override
+    public long offset()
+    {
+        return offset;
+    }
 
-        @VisibleForTesting
-        public void offset(long l)
+    @Override
+    public void release()
+    {
+        if (bufferHolder != null)
```

Comment:
We should be unforgiving to catch potential problems early. This doesn't work if we check `bufferHolder` as `TailOverridingRebufferer` leaves it null sometimes, but if we check and reset `buffer` instead (both here and in `rebuffer`), we can throw on double release too.